### PR TITLE
[Fix #3631] Add cop to check for spaces in '-> (' format lambdas

### LINKED ISSUE #3631: Add a cop to prevent `-> (` lambda format
As discovered in https://github.com/rmosolgo/graphql-ruby/pull/313:

- Both `-> (arg) { ... }` **and** `->(arg) { ... }` are supported in Ruby 2+ (notice the space between `->` and the parenthesis.
- `-> (arg) { ... }` is **not** supported in Ruby 1.9.3 and the interpreter reports:

```rb
syntax error, unexpected tLPAREN_ARG, expecting keyword_do_LAMBDA or tLAMBEG (SyntaxError)
      -> (arg) {
          ^
```

It seems reasonable to me to enforce `->(arg) { ... }` in style guides. After all, all examples in Rubocop's doc or your [Ruby style guide](https://github.com/bbatsov/ruby-style-guide) use this format consistently.

This is very valuable when developing a library that supports both versions 1.9.3 and 2+.

Looking forward to hearing your opinion on that.

### PR DESCRIPTION
**Pull request by swcraig**
I've created the cop with two configuration settings, require_space and require_no_space.
The default is require_no_space.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 * [#3408](https://github.com/bbatsov/rubocop/issues/3408): Add check for repeated values in case conditionals. ([@swcraig][])
 * [#3646](https://github.com/bbatsov/rubocop/pull/3646): Add new `Lint/EmptyWhen` cop. ([@drenmi][])
 * [#3246](https://github.com/bbatsov/rubocop/issues/3246): Add list of all cops to the manual (generated automatically from a rake task). ([@sihu][])
+* [#3631](https://github.com/bbatsov/rubocop/issues/3631): Add cop to check for spaces in `-> (` format lambdas. ([@swcraig][])
```
Comment (Drenmi) [inline CHANGELOG.md]: I find it useful to include the name of the cop in the `CHANGELOG`. 😀

File: lib/rubocop/cop/style/space_after_lambda_arrow.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+          receiver.nil? && call == :lambda
+        end
+
+        def arrow_form?(node)
+          node.loc.selector.source == ARROW
+        end
+
+        def args?(node)
+          _call, args, _body = *node.parent
+          !args.children.empty?
```
Comment (Drenmi) [inline lib/rubocop/cop/style/space_after_lambda_arrow.rb]: I'm not 100%, but I think `node.empty?` checks if there are any children.

**Comment by mikegee** [comment]
I agree with "lambda literal." `->` is special syntax that Ruby interprets like a call to `lambda`, just like quotes and brackets are special syntax for string, array, and hash literals that Ruby interprets as calls to `String.new`, etc.

**Comment by swcraig** [comment]
@bbatsov WDYT of this?

**Comment by bbatsov** [comment]
I merged this manually.

A small note - not sure whether this cop should warn for multiple spaces between `->` and `(`. Some people might be using them for alignment, but it seems like a bad idea to me.

**Comment by swcraig** [comment]
@bbatsov Thanks.
I am still relatively new to this workflow. In what cases would you manually merge a PR? Is it just when you would like to test the changes locally on your machine?

**Comment by bbatsov** [comment]
When there are conflicts in the changelog and I don't want to bother the authors to rebase.

**Comment by astorije** [comment]
Thanks for this! Any plans for a release yet? :)

> I don't want to bother the authors to rebase.

Btw, you can now edit the PR author branch (add commits, rebase existing commits) assuming the PR is set with edit rights to maintainers (it is by default).
It is roughly the same amount of steps, but with the appreciable perk that you don't have to close the PR.

**Pull request closed without merging**